### PR TITLE
[GEOS-7433] Add support for aggregation process group by results

### PR DESCRIPTION
**Pull request by nmco**
This pull request extends the current AggregatorProcess XML XML output format to support the new group by results. A new tabular JSON output format will be added. Backwards compatibility is granted.

Issue: https://osgeo-org.atlassian.net/projects/GEOS/issues/GEOS-7433?filter=myopenissues

### CODE REVIEW

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/AggregateProcessPPIO.java [hunk 1]
```diff
@@ -33,6 +37,10 @@ protected MapperWrapper wrapMapper(MapperWrapper next) {
             };
         };
         xstream.allowTypes(new Class[] { AggregateProcess.Results.class });
+        xstream.omitField(AggregateProcess.Results.class, "aggregateAttribute");
```
Comment (aaime) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/AggregateProcessPPIO.java]: So if someone uses the new group by functionality, the output will be available only in json?

Comment (aaime) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/AggregateProcessPPIO.java]: Btw, I have copied the omits and committed to get the build working again, but this is just a stop-gap measure,  imho if there is no grouping you should leave the extra fields null, or, as an alternative, write a custom xstream converter that will handle the grouping extra attributes only when grouping is actually used.

Comment (nmco) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/AggregateProcessPPIO.java]: I will put here some examples so we look both at the same output.

Simple WPS aggregate request without group by attributes:

XML
```
<?xml version="1.0" encoding="UTF-8"?>
<AggregationResults>
	<Max>500.0</Max>
</AggregationResults>
```

JSON
```
{
  "AggregationFunctions": ["Max"],
  "AggregationAttribute": "energy_consumption",
  "GroupByAttributes": [],
  "AggregationResults": [
    [500]
  ]
}
```

WPS request with group by attributes:

XML
```
<?xml version="1.0" encoding="UTF-8"?>
<AggregationResults>
	<GroupByResult>
		<object-array>
			<string>FABRIC</string>
			<big-decimal>500.0</big-decimal>
		</object-array>
		<object-array>
			<string>HOUSE</string>
			<big-decimal>6.0</big-decimal>
		</object-array>
		<object-array>
			<string>SCHOOL</string>
			<big-decimal>60.0</big-decimal>
		</object-array>
	</GroupByResult>
</AggregationResults>
```

JSON
```
{
  "AggregationFunctions": ["Max"],
  "AggregationAttribute": "energy_consumption",
  "GroupByAttributes": ["building_type"],
  "AggregationResults": [
    ["FABRIC",500],
    ["HOUSE",6],
    ["SCHOOL",60]
  ]
}
```

**Comment by jodygarnett** [comment]
looking at raw log shows the following failure:

```
[ERROR] Failed to execute goal on project gs-libjpeg-turbo: Could not resolve dependencies for project org.geoserver.extension:gs-libjpeg-turbo:jar:2.9-SNAPSHOT: Failed to collect dependencies at it.geosolutions.imageio-ext:imageio-ext-turbojpeg:jar:1.1.13 -> org.libjpegturbo:turbojpeg-wrapper:jar:1.2.1.1: Failed to read artifact descriptor for org.libjpegturbo:turbojpeg-wrapper:jar:1.2.1.1: Could not transfer artifact it.geosolutions.imageio-ext:imageio-ext-library:pom:1.2-SNAPSHOT from/to codehaus-snapshots (https://nexus.codehaus.org/snapshots/): nexus.codehaus.org: Unknown host nexus.codehaus.org -> [Help 1]
```

Does not seem to be related - odd that nexus.codehaus.org is referenced?

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/AggregateProcessPPIO.java (hunk 1 above)
Comment (jodygarnett) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/AggregateProcessPPIO.java]: Yes it is worth adding documentation, @bmmpxf is our usual contact point for docs.

Your examples are great, I think we could make a wps-cookbook of examples somewhere here:

* http://docs.geoserver.org/latest/en/user/services/wps/index.html

I would put it after the section on using the request builder, so you can use the request builder when making the examples/requests.

**Comment by jodygarnett** [comment]
merged to master, fixed header, organized docs (using the example your provided with your pull request). If you could kindly review, and update the example with a request based on the geoserver sample data it would be that much more useful.

**Comment by jodygarnett** [comment]
I tried figuring out how to run this process for examples:

image

Could not provide the "cat" group by values for 1,2,3 stream categories.

Please advise if you have a better group by example in mind?

**Comment by nmco** [comment]
Hello Jody,

Sorry for the delay (I hope this still be useful for you).

I send in attach two examples using geoserver shipped data (one defined in
the UI an another one using a XML).

By the way it seams that is not possible to provide several arguments in
the WPS builder (multiple aggregations functions, multiple group by
attributes).

In the XML example I do a count grouped by cat and label  on layer sf:roads
which gives this result:

{
  "AggregationFunctions": ["Count"],
  "AggregationAttribute": "cat",
  "GroupByAttributes": ["cat", "label"],
  "AggregationResults": [
    [5, "unimproved road",153],
    [3, "secondary highway, hard surface", 42],
    [2, "primary highway, hard surface", 60],
    [1, "interstate", 53],
    [4, "light-duty road, improved surface", 517]
  ]
}

So basically each road category as is own label.

Regarding the documentation, I think that WPS process cookbook is a very
good idea ... I don't have to much time now but I will certainly do a
pull request with some documentation.

Thank you for the review.



2016-02-25 3:07 GMT+00:00 Jody Garnett <notifications@github.com>:

**Comment by jodygarnett** [comment]
It totally is, will update the docs now

--
Jody Garnett

**Comment by nmco** [comment]
ooops sorry I took that github would keep the attachments, the XML query:

```
<?xml version="1.0" encoding="UTF-8"?><wps:Execute version="1.0.0" service="WPS" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://www.opengis.net/wps/1.0.0" xmlns:wfs="http://www.opengis.net/wfs" xmlns:wps="http://www.opengis.net/wps/1.0.0" xmlns:ows="http://www.opengis.net/ows/1.1" xmlns:gml="http://www.opengis.net/gml" xmlns:ogc="http://www.opengis.net/ogc" xmlns:wcs="http://www.opengis.net/wcs/1.1.1" xmlns:xlink="http://www.w3.org/1999/xlink" xsi:schemaLocation="http://www.opengis.net/wps/1.0.0 http://schemas.opengis.net/wps/1.0.0/wpsAll.xsd">
  <ows:Identifier>gs:Aggregate</ows:Identifier>
  <wps:DataInputs>
    <wps:Input>
      <ows:Identifier>features</ows:Identifier>
      <wps:Reference mimeType="text/xml" xlink:href="http://geoserver/wfs" method="POST">
        <wps:Body>
          <wfs:GetFeature service="WFS" version="1.0.0" outputFormat="GML2" xmlns:sf="http://www.openplans.org/spearfish">
            <wfs:Query typeName="sf:roads"/>
          </wfs:GetFeature>
        </wps:Body>
      </wps:Reference>
    </wps:Input>
    <wps:Input>
      <ows:Identifier>aggregationAttribute</ows:Identifier>
      <wps:Data>
        <wps:LiteralData>cat</wps:LiteralData>
      </wps:Data>
    </wps:Input>
    <wps:Input>
      <ows:Identifier>function</ows:Identifier>
      <wps:Data>
        <wps:LiteralData>Count</wps:LiteralData>
      </wps:Data>
    </wps:Input>
    <wps:Input>
      <ows:Identifier>singlePass</ows:Identifier>
      <wps:Data>
        <wps:LiteralData>false</wps:LiteralData>
      </wps:Data>
    </wps:Input>
    <wps:Input>
      <ows:Identifier>groupByAttributes</ows:Identifier>
      <wps:Data>
        <wps:LiteralData>cat</wps:LiteralData>
      </wps:Data>
    </wps:Input>
     <wps:Input>
      <ows:Identifier>groupByAttributes</ows:Identifier>
      <wps:Data>
        <wps:LiteralData>label</wps:LiteralData>
      </wps:Data>
    </wps:Input>
  </wps:DataInputs>
  <wps:ResponseForm>
    <wps:RawDataOutput mimeType="application/json">
      <ows:Identifier>result</ows:Identifier>
    </wps:RawDataOutput>
  </wps:ResponseForm>
</wps:Execute>
```

**Comment by nmco** [comment]
In the WPS request builder is not possible to give multiple group by attributes but here an example with one group by attribute:
wps-aggregate-example

**Comment by jodygarnett** [comment]
excellent, mind gathering that up into a documentation pull request. The page is here:

http://docs.geoserver.org/latest/en/user/services/wps/processes/gs.html#aggregation-process

**Comment by nmco** [comment]
Improved the WPS aggregate process documentation with real examples that
can be runned againts geoserver shipped data (if you want me to add the
previous examples or somethign else let me know):
https://github.com/geoserver/geoserver/pull/1519

2016-02-26 16:44 GMT+00:00 Jody Garnett <notifications@github.com>:

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/AggregateProcessPPIO.java (hunk 1 above)
Comment (Tony-yt-z) [reply src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/AggregateProcessPPIO.java:40]: > I will put here some examples so we look both at the same output (my comments and questions at the end). Note, this are the outputs produced by this pull request.
> 
> #### WPS aggregate request WITHOUT group by attributes:
> **_XML_**
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <AggregationResults>
>     <Max>500.0</Max>
> </AggregationResults>
> ```
> 
> **_JSON_**
> 
> ```
> {
>   "AggregationFunctions": ["Max"],
>   "AggregationAttribute": "energy_consumption",
>   "GroupByAttributes": [],
>   "AggregationResults": [
>     [500]
>   ]
> }
> ```
> 
> #### WPS aggregate request WITH group by attributes:
> **_XML_**
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <AggregationResults>
>     <GroupByResult>
>         <object-array>
>             <string>FABRIC</string>
>             <big-decimal>500.0</big-decimal>
>         </object-array>
>         <object-array>
>             <string>HOUSE</string>
>             <big-decimal>6.0</big-decimal>
>         </object-array>
>         <object-array>
>             <string>SCHOOL</string>
>             <big-decimal>60.0</big-decimal>
>         </object-array>
>     </GroupByResult>
> </AggregationResults>
> ```
> 
> **_JSON_**
> 
> ```
> {
>   "AggregationFunctions": ["Max"],
>   "AggregationAttribute": "energy_consumption",
>   "GroupByAttributes": ["building_type"],
>   "AggregationResults": [
>     ["FABRIC",500],
>     ["HOUSE",6],
>     ["SCHOOL",60]
>   ]
> }
> ```
> 
> #### Comments
> * Answering your question: yes the current XML output can handle the group by result and is backwards compatible if there is no group by results (at least this was my intention).
> * Do you think is worth to add headers info to the XML output like in the JSON output (probably yes) ? The results are returned in the order of the invocations (group by results and functions this is also the case in the JSON format).
> * Do you think is worth add a documentation page for this process ? If yes where should I create a page in GeoServer documentation for this ?
> * Regarding the omit fields, I was only omitting fields I had to add to the process result so I had access to some extra information when producing the JSON output (to create the headers) since I don't produce headers in the XML output format I was just ignoring that information.
> * Looking at the examples if you don't agree with something please let me know.
> 
> The next time I will put better examples on my pull request description and thanks for the review.



> I will put here some examples so we look both at the same output (my comments and questions at the end). Note, this are the outputs produced by this pull request.
> 
> #### WPS aggregate request WITHOUT group by attributes:
> **_XML_**
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <AggregationResults>
>     <Max>500.0</Max>
> </AggregationResults>
> ```
> 
> **_JSON_**
> 
> ```
> {
>   "AggregationFunctions": ["Max"],
>   "AggregationAttribute": "energy_consumption",
>   "GroupByAttributes": [],
>   "AggregationResults": [
>     [500]
>   ]
> }
> ```
> 
> #### WPS aggregate request WITH group by attributes:
> **_XML_**
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <AggregationResults>
>     <GroupByResult>
>         <object-array>
>             <string>FABRIC</string>
>             <big-decimal>500.0</big-decimal>
>         </object-array>
>         <object-array>
>             <string>HOUSE</string>
>             <big-decimal>6.0</big-decimal>
>         </object-array>
>         <object-array>
>             <string>SCHOOL</string>
>             <big-decimal>60.0</big-decimal>
>         </object-array>
>     </GroupByResult>
> </AggregationResults>
> ```
> 
> **_JSON_**
> 
> ```
> {
>   "AggregationFunctions": ["Max"],
>   "AggregationAttribute": "energy_consumption",
>   "GroupByAttributes": ["building_type"],
>   "AggregationResults": [
>     ["FABRIC",500],
>     ["HOUSE",6],
>     ["SCHOOL",60]
>   ]
> }
> ```
> 
> #### Comments
> * Answering your question: yes the current XML output can handle the group by result and is backwards compatible if there is no group by results (at least this was my intention).
> * Do you think is worth to add headers info to the XML output like in the JSON output (probably yes) ? The results are returned in the order of the invocations (group by results and functions this is also the case in the JSON format).
> * Do you think is worth add a documentation page for this process ? If yes where should I create a page in GeoServer documentation for this ?
> * Regarding the omit fields, I was only omitting fields I had to add to the process result so I had access to some extra information when producing the JSON output (to create the headers) since I don't produce headers in the XML output format I was just ignoring that information.
> * Looking at the examples if you don't agree with something please let me know.
> 
> The next time I will put better examples on my pull request description and thanks for the review.

So, when WPS aggregate request WITH group by attributes,and output xml results, can I get the output in the order I entered it?

Comment (Tony-yt-z) [reply src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/AggregateProcessPPIO.java:40]: and when i output json format, it doesn't seem to support Chinese

Comment (Tony-yt-z) [reply src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/AggregateProcessPPIO.java:40]: @nmco

Comment (Tony-yt-z) [reply src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/AggregateProcessPPIO.java:40]: 微信图片_20230309141331
